blueprints/distributed-traffic-monitor.md

```diff
@@ -0,0 +1,180 @@
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+# Distributed Traffic Monitor
+
+## Problem Description
+Currently, TM polls all caches in a CDN. As CDNs grow, this becomes a major
+pain point as TM is limited by the amount of bandwidth and CPU it requires to
+receive and process data from every cache on the CDN, and scaling vertically by
+running it on better hardware is only feasible up to a certain point. Also, the
+performance of a cache observed by a TM which is very far away from it does not
+always reflect the performance observed by clients that are actually using the
+cache (because the clients are typically much closer to it).
+
+## Proposed Change
+TM should have the ability to poll only a subset of caches in a CDN and peer
+with other TMs which are monitoring other subsets in order to get a full view
+of the CDN's health. This would allow us to run TM in a more distributed manner
+across the CDN, giving us a view of cache health that is closer to what clients
+actually observe and enabling us to scale TM horizontally. Additionally, we
+would like to have the option to disable _stat polling_ in order for these
+distributed TMs to focus on _health polling_.
+
+### Traffic Portal Impact
+This proposal does not require any TP changes.
+
+### Traffic Ops Impact
+This proposal should have no impact on TO. The existing TO API endpoints
+already provide the data that TM will need to run in a distributed manner.
```

Comment:
No, in the TM impact section I mention that there will be a _new_ TM API that only returns health states for a subset of the caches in a CDN, but I can mention that any changes to existing TM APIs will be backwards-compatible so that any TO APIs that use them will be unaffected.

Reply:
I'm sure the old endpoints will _structurally_ remain the same, the problem is that you can't get all the health data for a CDN from just one TM anymore, which is currently how it's implemented

Reply:
You will be able to get the _health_ data for a CDN from any TM, but you won't be able to get all the _stats_ data for a CDN from any TM.

Reply:
TMs will store health information for caches they don't poll for health? I think the blueprint says they will be getting them from peer-polling anyway, I suppose.

Reply:
Yes, it will store health information for all caches in the CDN. It has to, because TR expects to be able to talk to any TM and get all the cache health states. That is why in distributed mode it will still have to query TMs from other "TM groups."